engine/schema/src/main/java/com/cloud/host/dao/HostDaoImpl.java

```diff
@@ -1756,6 +1758,29 @@ public List<HypervisorType> listDistinctHypervisorTypes(final Long zoneId) {
         return customSearch(sc, null);
     }
 
+    @Override
+    public List<Pair<HypervisorType, CPU.CPUArch>> listDistinctHypervisorArchTypes(final Long zoneId) {
+        List<Pair<HypervisorType, CPU.CPUArch>> hypervisorArchList = new ArrayList<>();
+        String selectSql = "SELECT DISTINCT hypervisor_type, arch FROM cloud.host WHERE removed IS NULL";
+        if (zoneId != null) {
+            selectSql += " AND data_center_id=" + zoneId;
+        }
+        TransactionLegacy txn = TransactionLegacy.currentTxn();
+        try {
+            PreparedStatement stmt = txn.prepareAutoCloseStatement(selectSql);
+            ResultSet rs = stmt.executeQuery();
+            while (rs.next()) {
+                HypervisorType hypervisorType = HypervisorType.valueOf(rs.getString("hypervisor_type"));
+                CPU.CPUArch arch = CPU.CPUArch.fromType(rs.getString("arch"));
+                hypervisorArchList.add(new Pair<>(hypervisorType, arch));
+            }
+        } catch (SQLException ex) {
+            logger.error("DB exception {}", ex.getMessage(), ex);
+            return Collections.emptyList();
+        }
+        return hypervisorArchList;
+    }
+
```

Comment:
If possible can we include this method on `GenericDaoBase` accepting a table as parameter so then it can be invoked on ClusterDaoImpl and here just passing the different table to query from?

Reply:
@nvazquez moving a method with specific columns in GenericDaoBase doesn't sound right to me.
I've refactored the code use SearchBuilder and SearchCriteria instead.